Mathlib/Data/Matroid/Basic.lean

```diff
@@ -0,0 +1,884 @@
+
+[2] The robust axiomatic definition of infinite matroids
+[H. Bruhn, R. Diestel, M. Kriesell, R. Pendavingh, P. Wollan, Axioms for infinite matroids,
+  Adv. Math 239 (2013), 18-46]
+
+[3] Equicardinality of matroid bases is independent of ZFC.
+[N. Bowler, S. Geschke, Self-dual uniform matroids on infinite sets,
+  Proc. Amer. Math. Soc. 144 (2016), 459-471]
+-/
+
+
+
+open Set
+
+/-- A predicate `P` on sets satisfies the exchange property if, for all `X` and `Y` satisfying `P`
+  and all `a ∈ X \ Y`, there exists `b ∈ Y \ X` so that swapping `a` for `b` in `X` maintains `P`.-/
+def Matroid.ExchangeProperty {α : Type _} (P : Set α → Prop) : Prop :=
+  ∀ X Y, P X → P Y → ∀ a ∈ X \ Y, ∃ b ∈ Y \ X, P (insert b (X \ {a}))
+
+/-- A set `X` has the maximal subset property for a predicate `P` if every subset of `X` satisfying
+  `P` is contained in a maximal subset of `X` satisfying `P`.  -/
+def Matroid.ExistsMaximalSubsetProperty {α : Type _} (P : Set α → Prop) (X : Set α) : Prop :=
+  ∀ I, P I → I ⊆ X → (maximals (· ⊆ ·) {Y | P Y ∧ I ⊆ Y ∧ Y ⊆ X}).Nonempty
+
+/-- A `Matroid α` is a `ground` set of type `Set α`, and a nonempty collection of its subsets
+  satisfying the exchange property and the maximal subset property. Each such set is called a
+  `Base` of `M`. -/
+@[ext] structure Matroid (α : Type _) where
+  /-- `M` has a ground set. -/
+  (ground : Set α)
+  /-- `M` has a predicate `Base` definining its bases -/
+  (Base : Set α → Prop)
+  /-- There is at least one `Base` -/
+  (exists_base' : ∃ B, Base B)
+  /-- For any bases `B,B'` and `e ∈ B \ B'`, there is some `f ∈ B' \ B` for which `B-e+f`
+    is a base -/
+  (base_exchange' : Matroid.ExchangeProperty Base)
+  /-- Every subset `I` of a set `X` for which `I` is contained in a base is contained in a maximal
+    subset of `X` that is contained in a base. -/
+  (maximality' : ∀ X, X ⊆ ground → Matroid.ExistsMaximalSubsetProperty (∃ B, Base B ∧ · ⊆ B) X)
+  /-- every base is contained in the ground set -/
+  (subset_ground' : ∀ B, Base B → B ⊆ ground)
+
+namespace Matroid
+
+variable {α : Type _} {M : Matroid α}
+
+attribute [pp_dot] Base
+
+/-- We write `M.E` for the ground set of a matroid `M`-/
+@[pp_dot] def E (M : Matroid α) : Set α := M.ground
+
+@[simp] theorem ground_eq_E (M : Matroid α) : M.ground = M.E := rfl
+
+/-- Typeclass for a matroid having finite ground set. Just a wrapper for `M.E.Finite`-/
+class Finite (M : Matroid α) : Prop where
+  /-- The ground set is finite -/
+  (ground_finite : M.E.Finite)
+
+/-- Typeclass for a matroid having nonempty ground set. Just a wrapper for `M.E.Nonempty`-/
+class Nonempty (M : Matroid α) : Prop where
+  /-- The ground set is nonempty -/
+  (ground_nonempty : M.E.Nonempty)
+
+theorem ground_nonempty (M : Matroid α) [Nonempty M] : M.E.Nonempty :=
+  Nonempty.ground_nonempty
+
+theorem ground_finite (M : Matroid α) [Finite M] : M.E.Finite :=
+  ‹M.Finite›.ground_finite
```

Comment:
I think that you can make these instances in Lean 4: loops in the typeclass hierarchy are no longer a problem.

Reply:
Is it normal to have `Set.Finite` terms as instances?

Reply:
Ooh, sorry. I forgot that this was about `Set` instead of about types. So your version is correct.